New Feature: SNOMED::ICD10CM Mapping Support

joeflack4 opened the pull request:
```
New Feature: SNOMED::ICD10CM Mapping Support
- Added feature to allow for conversion of these premade mappings provided by SNOMED into SSSOM format. (WIP)

Dev updates
- requirements-unlocked.txt: Unversioned variation of requirements.txt
- requirements.txt: Included this because I don't know if setup.py is sufficient to install packages in a standalone dev environment. Also, it is shown to be expected in the documentation: https://mapping-commons.github.io/sssom-py/installation.html#installation-for-developers
- ```

matentzn commented:
The actual extension is looking great! @hrshdhgd can take it for a ride when the time comes.

Sorry we have not introduced you to our build system: we are using `tox` - please do not commit any requirements.txts here. To build sssom-py, just use `tox`. If you have questions on that @hrshdhgd will be happy to help!

joeflack4 commented:
I updated my draft PR:
https://github.com/mapping-commons/sssom-py/pull/207

I think I'm almost done with the core of it. I'll upload an example output tonight/tomorrow. There are some fields that I'm not too sure on. I've added a lot of comments here:
https://github.com/mapping-commons/sssom-py/blob/snomed-icd-mappings-work/sssom/parsers.py#L520

joeflack4 commented on sssom/parsers.py:
```diff
@@ -499,6 +518,144 @@ def from_obographs(
+            #   SNOMED mappings, all of them had correlationId of 447561005, which also happens to be 'unspecified'.
+            #   If correlationId is indeed more appropriate for predicate_id, then I don't think there is a representative
+            #   field for 'match_type'.
+            'match_type': MatchTypeEnum('Unspecified') if row['correlationId'] == match_type_snomed_unspecified_id \
```
### `match_type` from `correlationId`
Everything was unspecified. There appears to be a SNOMED code for that, so we could put `SNOMED: 447561005` here instead of `'Unspecified'`.

matentzn replied:
Interesting. Never thought of that! Maybe good. Need a quick discussion.

joeflack4 replied:
Sounds good. I added to the agenda in my notes for our next meeting.

joeflack4 commented on sssom/parsers.py:
```diff
@@ -499,6 +518,144 @@ def from_obographs(
+            'object_label': row['mapTargetName'],
+
+            # match_type <- mapRule?
+            #   ex: TRUE: when "ALWAYS <code>" is in pipe-delimited list in mapAdvice, this always shows TRUE. Does this
```
### `match_type` or `predicate_id` from `mapRule`
@matentzn This in particular I might want to go back and change. The `mapRule` column is a `bool`. I _think_ it may be the case that all instances of `TRUE` in this column could indicate that they are exact matches, because this field only shows `TRUE` when the `mapAdvice` field includes text that says `ALWAYS <code>`.

But by "ALWAYS" and "TRUE", I don't know if that means that (a) the mapped SNOMED::ICD code are the only codes that are related to each other, or (b) it's an exact match, in which case we could use `skos:exactMatch` here for that (also, does SKOS not have integer IDs for its concepts?).

I'm thinking that (b) is more likely the case, but I'm not too sure. Maybe someone with more domain expertise could look at some of the example cases. As mentioned [here](https://www.nlm.nih.gov/research/umls/mapping_projects/snomedct_to_icd10cm.html), the map TSV can be downloaded here:
> The map is also available in a human readable format (tls_Icd10cmHumanReadableMap_US1000124_YYYYMMDD.tsv) and can be downloaded from the [US Edition page](https://www.nlm.nih.gov/healthit/snomedct/us_edition.html).

matentzn replied:
Let's do this in our next 1:1 rather than discuss it in this PR. You are asking all the right questions, but answers are complicated and its better done in person.

joeflack4 replied:
Sounds good. added to my agenda notes

joeflack4 replied:
Sounds good. added my agenda

joeflack4 commented on sssom/parsers.py:
```diff
@@ -499,6 +518,144 @@ def from_obographs(
+            #   Example: Not Negates the predicate, see documentation of predicate_modifier_enum
+            # - predicate_id <- mapAdvice?
+            # - predicate_modifier <- mapAdvice?
+            #   mapAdvice: Pipe-delimited qualifiers. Ex:
```
### `predicate_modifier` from `mapAdvice`
The `mapAdvice` field is a pipe-delimited list of advice. If there is an `ALWAYS <code>` present, I think it's always the first thing to show up in the list. I don't think that particular bit of information is useful for determining `predicate_modifier`, but I wonder if maybe some other information present in this `mapAdvice` field could sometimes be useful for this? I'm not sure what all the `predicate_modifiers` are that are available. I only saw the "Not" modifier in the docs.

matentzn replied:
SSSOM _only_ cares about `ALWAYS`. So you should filter the incoming data by that and add a line to the documentation (see above) to mention that decision. There is no notion in SSSOM for mapAdvice - this is a complicated contextual parameter, which we are ignoring for now.

joeflack4 replied:
Done. There are some caveats / questions / edge cases though. Will flag this in another review comment shortly.

joeflack4 commented on sssom/parsers.py:
```diff
@@ -499,6 +518,144 @@ def from_obographs(
+            #   or is there a SKOS predicate I can map to in case where predicate is unknown? I think most of these
+            #   mappings are attempts at exact matches, but I can't be sure (at least not without using these fields
+            #   to determine: mapGroup, mapPriority, mapRule, mapAdvice).
+            # mapCategoryId,mapCategoryName: Only these in set: 447637006 "MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED",
```
### `mapCategoryId` and `mapCategoryName`
I'm not sure if there's any useful information here that could be put anywhere than the `other` field. There are only 3 values that show up in the entire mapping dataset for both `mapCategoryId` and `mapCategoryName`, and I listed them out int he comment here.

hrshdhgd replied:
@matentzn

joeflack4 commented on sssom/parsers.py:
```diff
@@ -499,6 +518,144 @@ def from_obographs(
+            #   "IF LISSENCEPHALY TYPE 3 FAMILIAL FETAL AKINESIA SEQUENCE SYNDROME CHOOSE Q04.3 | MAP OF SOURCE CONCEPT
+            #   IS CONTEXT DEPENDENT"
+            #   "MAP SOURCE CONCEPT CANNOT BE CLASSIFIED WITH AVAILABLE DATA"
+            'predicate_id': f'SNOMED:{row["mapCategoryId"]}',
```
As I mentioned elsewhere, I'm not 100% sure if `predicate_*` is best taken from `mapCategory*`. It's possible that `mapRule` and `mapAdvice` could be better. Would like to defer to others for judgement on this.

hrshdhgd replied:
@matentzn

matentzn replied:
I think a much deeper understanding of these is needed.. Ask in the call later.

matentzn commented on sssom/parsers.py:
```diff
@@ -140,6 +141,24 @@ def read_obographs_json(
     )
 
 
+def read_snomed_icd10cm_map_tsv(
```
Great!

matentzn commented on sssom/parsers.py:
```diff
@@ -140,6 +141,24 @@ def read_obographs_json(
+) -> MappingSetDataFrame:
+    """Parse special SNOMED ICD10CM mapping file and translates it into a MappingSetDataFrame.
+
+    :param file_path: The path to the obographs file
```
path to icd10 mapping file, see https://www.nlm.nih.gov/research/umls/mapping_projects/snomedct_to_icd10cm.html

matentzn replied:
Actually I wonder if that file format is something more general - i.e. snomed mapping. There are many more such files with other kinds of mappings in them.. maybe rename methods to "snomed_mapping" rather than "snomed_icd10_mapping"..

joeflack4 replied:
Good find! Done! Also, I determined that this is a 'complex' mapping based on what you linked earlier.

matentzn commented on sssom/parsers.py:
```diff
@@ -499,6 +518,144 @@ def from_obographs(
+    :param meta: A metadata dictionary
+    :return: MappingSetDataFrame
+
+    # Field descriptions
```
Can we add a new file, `sphinx/parsers.rst` and add a section in the SNOMED mapping to SSSOM? Dont worry about the details, you can just for now cut and paste everything there and link it from the code.

joeflack4 replied:
Good idea. I'm pretty terrible at `.rst` syntax, so I mostly just copy/pasted as you said.

matentzn commented on sssom/parsers.py:
```diff
@@ -499,6 +518,144 @@ def from_obographs(
+                else  MatchTypeEnum('Unspecified'),
+
+            'mapping_date': date_parser.parse(str(row['effectiveTime'])).date(),
+            'other': '|'.join([f'{k}={str(row[k])}' for k in [
```
Great thanks!

matentzn commented on sssom/parsers.py:
```diff
@@ -499,6 +518,144 @@ def from_obographs(
+                'mapAdvice',
+            ]]),
+
+            # More fields (https://mapping-commons.github.io/sssom/Mapping/):
```
Lets do this in person.

joeflack4 replied:
Roger that.

matentzn reviewed (commented):
I am super excited about this, and I am very happy about how much care you are taking sorting out the mapping here. Most of the things in here we need to discuss in person, too much to unpack to type up now, but two things would be probably good:

1. Start a new "parser" documentation as described in comments above
2. Rebrand the method to "snomed_map" rather than "snomed_icd10_map". In fact, there are two very different ones: complex and simple map. So lets figure out which of the two you are dealing with at the moment, see also: https://github.com/mapping-commons/sssom/issues/42 and rebrand your methods accordingly. So `read_snomed_simple_map()` or `read_snomed_complex_map()` depending on what we are dealing with here.

joeflack4 commented on sphinx/parsers.rst:
```diff
@@ -0,0 +1,126 @@
+Parsers
```
I'm pretty terrible at `.rst`, but here you go. Based on what you said, this should be suitable for right now.

joeflack4 commented on sssom/parsers.py:
```diff
@@ -140,6 +141,28 @@ def read_obographs_json(
+    file_path: str,
+    prefix_map: Dict[str, str] = None,
+    meta: Dict[str, str] = None,
+    filter_by_confident_mappings=True
```
Based on your request to filter only mapping rows that contain `'ALWAYS <code>'`, and possible future flexibility to fine tune the mappings, I added a param here and just set it to `True`.

hrshdhgd replied:
@matentzn

joeflack4 commented on sssom/parsers.py:
```diff
@@ -499,6 +522,159 @@ def from_obographs(
+    # - Note: joeflack4: I used this info as a reference for this pattern.
+    # https://www.medicalbillingandcoding.org/icd-10-cm/#:~:text=ICD%2D10%2DCM%20is%20a,decimal%20point%20and%20the%20subcategory.
+    always_confidence_pattern = 'ALWAYS [A-Z]{1}[0-9]{1,2}\.[0-9A-Z]{1,4}'
+    always_confidence_antipattern = always_confidence_pattern + '\?'
```
I found edge cases where `'ALWAYS <code>'` appears, but has a `?` mark after, indicating that it is not 100% confident. Example:

> ALWAYS T50.905? | CONSIDER ADDITIONAL CODE TO IDENTIFY SPECIFIC CONDITION OR DISEASE | EPISODE OF CARE INFORMATION NEEDED

hrshdhgd replied:
@matentzn

joeflack4 commented on sssom/parsers.py:
```diff
@@ -499,6 +522,159 @@ def from_obographs(
+    ms = _init_mapping_set(meta)
+
+    # Filtering
+    if filter_by_confident_mappings:
```
Here's the filtering you requested.

It turns out that I was wrong earlier when I thought that if `mapAdvice` contained `'ALWAYS <code>'`, `mapRule` would always be `TRUE`. If this was the case, filtering by `mapRule` would've been much easier, but unfortunately it did not turn out to be the case.

I was thinking about a simple `.str.contains('ALWAYS')`, but it is possible that the word 'ALWAYS' could appear in cases other than `'ALWAYS <code>'`, so I decided to go with the stricter approach of a regex.

I created my own regex. But now I just realized that ICD10 is such a big thing that I should've just been googled a pre-baked regex. Here's one I found: https://www.johndcook.com/blog/2019/05/05/regex_icd_codes/. If you'd like me to use that or another regex instead, let me know.

I checked to see how many rows were removed by this filter, and it was about ~50%.

hrshdhgd replied:
@matentzn

joeflack4 commented on sssom/parsers.py:
```diff
@@ -19,9 +20,11 @@
 from rdflib import Graph, URIRef
 
-# from .sssom_datamodel import Mapping, MappingSet
+# TODO: PR comment: where matchtypeenum? can't find sssomschema, Mapping, or MappingSet. only MappingSetDataFrame
```
@matentzn I rebased from `master`, but I have 2 issues that need fixing that I need help with:
## Issue 1: `ModuleNotFoundError: No module named 'sssom_schema'`
I saw that what I was previously importing from `.sssom_datamodel` is now imported from `sssom_schema`. I assume this comes with a dependency. So I ran `tox` to install the latest dependencies, but even the test suite failed on this very issue. Here's the full output from `.tox`:

[tox-output.txt](https://github.com/mapping-commons/sssom-py/files/9105522/tox-output.txt)

## Issue 2: Where is `MatchTypeEnum`?
I found `MatchTypeEnum` useful and I want it back. I'm assuming this is now in `sssom_schema`, but I need to resolve issue (1) before I can get it. Or, if it's not there, I can add MatchTypeEnum back somewhere. Here is the definition that I pulled from a previous commit:

```py
class MatchTypeEnum(EnumDefinitionImpl):

    Lexical = PermissibleValue(text="Lexical", description="Lexical match")
    Logical = PermissibleValue(text="Logical", description="Logical match")
    HumanCurated = PermissibleValue(
        text="HumanCurated", description="Match based on human expert opinion"
    )
    Complex = PermissibleValue(
        text="Complex", description="Match based on a variety of different strategies"
    )
    Unspecified = PermissibleValue(text="Unspecified", description="Unknown match type")
    SemanticSimilarity = PermissibleValue(
        text="SemanticSimilarity",
        description="Match based on close semantic similarity",
    )

    _defn = EnumDefinition(
        name="MatchTypeEnum",
    )
```

hrshdhgd replied:
Nico, please correct me if I'm wrong.
Joe:
First check: Have you pulled from the main branch?
Issue 1: replace `from .sssom_datamodel import Mapping, MappingSet` to `from sssom_schema import Mapping, MappingSet` I think this should fix it.

Issue 2: `MatchTypeEnum` is now replace by [SEMAPV](https://github.com/mapping-commons/sssom-py/blob/da1256830f4f2dbe4eb70fbe78945074e882ca52/sssom/constants.py#L122). 

I hope this helps!

joeflack4 replied:
I got tox working now, thanks. For issue 2, I am going to close this and open a new comment.

joeflack4 replied:
@hrshdhgd Thanks so much for responding to this!

Issue 1
Yeah I did get the latest code. I did a `rebase` of my branch from `master`. I pinged Nico and he asked me to just remove my `.tox` dir and run the command again. that worked.

Issue 2
Very helpful, thanks! This is exactly what I was looking for. I tried various searches to see where this was moved to but couldn't find it; search fail on my part.

I will update my code to use this, though I think the only SNOMED::ICD mapping that I was able to ascertain from the  mapping file that SNOMED published was "UnspecifiedMatching", but I should probably double check.

joeflack4 commented on sssom/parsers.py:
```diff
@@ -666,6 +689,163 @@ def from_obographs(
+            #   SNOMED mappings, all of them had correlationId of 447561005, which also happens to be 'unspecified'.
+            #   If correlationId is indeed more appropriate for predicate_id, then I don't think there is a representative
+            #   field for 'mapping_justification'.
+            # TODO: How to properly get mapping_justification?
```
## Issue 1
Not sure how to use `sssom_schema.slots.mapping_justification` here.

## Issue 2
Could take some time to map all the SNOMED mapping fields to SSSOM mapping fields.

---

@matentzn I recommend we merge this now while this is rebased and functional, and we can make improvements to this feature later.

hrshdhgd commented:
I did some formatting Joe, I hope that was fine. The pending errors from `tox` are as follows:

```
sssom/parsers.py:273:1: DAR003 Incorrect indentation: ~<
sssom/parsers.py:704:1: DAR003 Incorrect indentation: ~<
sssom/parsers.py:745:59: W605 invalid escape sequence '\.'
sssom/parsers.py:746:66: W605 invalid escape sequence '\?'
```

hrshdhgd commented:
Fixed the errors: 
sssom/parsers.py:745:59: W605 invalid escape sequence '\.'
sssom/parsers.py:746:66: W605 invalid escape sequence '\?'

by adding a `r` before the regex expression. Still the other 2 are pending.

matentzn commented on docs/parsers.rst:
```diff
@@ -0,0 +1,126 @@
+#   IS CONTEXT DEPENDENT"
+#   "MAP SOURCE CONCEPT CANNOT BE CLASSIFIED WITH AVAILABLE DATA"
+'predicate_id': f'SNOMED:{row["mapCategoryId"]}',
+'predicate_label': row['mapCategoryName'],
```
Can you share an example of how the resulting sssom looks like, maybe, 10 rows in markdown here in this pull request?

joeflack4 replied:
Yep, sure thing.

joeflack4 replied:
|subject_id            |predicate_id    |object_id      |match_type |subject_label                               |subject_category|predicate_label                           |predicate_modifier|object_label                                        |object_category|author_id|author_label|reviewer_id|reviewer_label|creator_id|creator_label|license|subject_source|subject_source_version|object_source|object_source_version|mapping_provider|mapping_cardinality|mapping_tool|mapping_tool_version|mapping_date|confidence|subject_match_field|object_match_field|match_string|subject_preprocessing|object_preprocessing|match_term_type|semantic_similarity_score|semantic_similarity_measure|see_also|other                                                                                                                                                                                                                                                                                                                                                                                                                                                 |comment|
|----------------------|----------------|---------------|-----------|--------------------------------------------|----------------|------------------------------------------|------------------|----------------------------------------------------|---------------|---------|------------|-----------|--------------|----------|-------------|-------|--------------|----------------------|-------------|---------------------|----------------|-------------------|------------|--------------------|------------|----------|-------------------|------------------|------------|---------------------|--------------------|---------------|-------------------------|---------------------------|--------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------|
|SNOMED:10000006       |SNOMED:447637006|ICD10CM:R07.9  |Unspecified|Radiating chest pain                        |                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Chest pain, unspecified                             |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=4c4d2755-e65f-52dc-8353-ecf437495eb8&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=TRUE&#124;mapAdvice=ALWAYS R07.9                                                                                                                                                                                                                                                                                           |       |
|SNOMED:10001005       |SNOMED:447639009|ICD10CM:P36.9  |Unspecified|Bacterial sepsis                            |                |MAP OF SOURCE CONCEPT IS CONTEXT DEPENDENT|                  |Bacterial sepsis of newborn, unspecified            |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=f7e1ed02-3107-5a6c-9439-21e9879fb746&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=IFA 445518008 &#124; Age at onset of clinical finding (observable entity) &#124; < 29.0 days&#124;mapAdvice=IF AGE AT ONSET OF CLINICAL FINDING BEFORE 29.0 DAYS CHOOSE P36.9 &#124; CONSIDER ADDITIONAL CODE TO IDENTIFY SPECIFIC CONDITION OR DISEASE &#124; DESCENDANTS NOT EXHAUSTIVELY MAPPED &#124; MAP OF SOURCE CONCEPT IS CONTEXT DEPENDENT|       |
|SNOMED:10001005       |SNOMED:447637006|ICD10CM:A41.9  |Unspecified|Bacterial sepsis                            |                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Sepsis, unspecified organism                        |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=907aa2b0-a832-5254-acdf-6f366e406d2b&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=2&#124;mapRule=OTHERWISE TRUE&#124;mapAdvice=ALWAYS A41.9 &#124; CONSIDER ADDITIONAL CODE TO IDENTIFY SPECIFIC CONDITION OR DISEASE &#124; DESCENDANTS NOT EXHAUSTIVELY MAPPED                                                                                                                                                                      |       |
|SNOMED:10007009       |SNOMED:447637006|ICD10CM:Q04.8  |Unspecified|Coffin-Siris syndrome                       |                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Other specified congenital malformations of brain   |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=0b918513-be22-5059-beb6-e8348dfc631f&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=TRUE&#124;mapAdvice=ALWAYS Q04.8                                                                                                                                                                                                                                                                                           |       |
|SNOMED:1001000119102  |SNOMED:447637006|ICD10CM:I26.99 |Unspecified|Pulmonary embolism with pulmonary infarction|                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Other pulmonary embolism without acute cor pulmonale|               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=7f3f0a95-0441-57ba-a58f-03d2f2736c60&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=TRUE&#124;mapAdvice=ALWAYS I26.99                                                                                                                                                                                                                                                                                          |       |
|SNOMED:10017004       |SNOMED:447637006|ICD10CM:K03.0  |Unspecified|Occlusal wear of teeth                      |                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Excessive attrition of teeth                        |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=3fa74c85-56d0-5819-97e8-80c8d4c053c0&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=TRUE&#124;mapAdvice=ALWAYS K03.0                                                                                                                                                                                                                                                                                           |       |
|SNOMED:100191000119105|SNOMED:447637006|ICD10CM:N42.89 |Unspecified|Asymmetry of prostate                       |                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Other specified disorders of prostate               |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=51537076-a3bf-59f8-aa5f-f571dfcf32cb&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=TRUE&#124;mapAdvice=ALWAYS N42.89                                                                                                                                                                                                                                                                                          |       |
|SNOMED:100211000119106|SNOMED:447637006|ICD10CM:M62.830|Unspecified|Muscle spasm of thoracic back               |                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Muscle spasm of back                                |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=a0263a65-4056-5cf9-8caa-669afce20189&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=TRUE&#124;mapAdvice=ALWAYS M62.830                                                                                                                                                                                                                                                                                         |       |
|SNOMED:1002229008     |SNOMED:447637006|ICD10CM:L75.2  |Unspecified|Apocrine miliaria of areola                 |                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Apocrine miliaria                                   |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=65fa44e5-272d-54e5-b787-cb977fc24720&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=TRUE&#124;mapAdvice=ALWAYS L75.2                                                                                                                                                                                                                                                                                           |       |
|SNOMED:1002253002     |SNOMED:447637006|ICD10CM:Z28.3  |Unspecified|Immunization series incomplete              |                |MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED |                  |Underimmunization status                            |               |         |            |           |              |          |             |       |              |                      |             |                     |                |                   |            |                    |2021-09-01  |          |                   |                  |            |                     |                    |               |                         |                           |        |id=c2ab86f4-29c5-5ea3-8818-1e389d9c59a6&#124;active=1&#124;moduleId=5991000124107&#124;refsetId=6011000124106&#124;mapGroup=1&#124;mapPriority=1&#124;mapRule=TRUE&#124;mapAdvice=ALWAYS Z28.3 &#124; CONSIDER ADDITIONAL CODE TO IDENTIFY SPECIFIC CONDITION OR DISEASE                                                                                                                                                                                                                      |       |

matentzn replied:
- match_type does not exist anymore
- Why all these empty columns?
- predicate_id is not legal - I looked this SNOMED rel up but could not understand what it means at all. Can you explain how "Map source concept is properly classified" is a mapping relation? how does it map to skos?
- you should add a test that runs `validate_file` on the a small sample of the output I think,

joeflack4 replied:
* `match_type`: Oh this is an old copy. I haven't re-ran it in awhile. The code has since been updated to `mapping_justification`.
* Empty columns: I haven't determined how to populate these SSSOM fields. For many of them, there may not be any information available. I have enumerated the SSSOM Mapping data model as of 2022/30 in my code comments ([link](https://github.com/mapping-commons/sssom-py/blob/snomed-icd-mappings-work/sssom/parsers.py#L825)). If you click that link and check it out, you'll see that I have some thoughts/questions as to how to populate. Any feedback/suggestions to those questions/comments are appreciated.
* `predicate_id`: I opened a [review comment](https://github.com/mapping-commons/sssom-py/pull/207#discussion_r819104571) in the first commit about populating this field from SNOMED's `mapCategoryId` and `mapCategoryName` fields. They don't use too many mapping predicates, unfortunately. In that linked review comment, I stated that there were only 3 values in the field, but the ICD10CM mapping dataset only includes 1: "MAP SOURCE CONCEPT IS PROPERLY CLASSIFIED" (447637006). You might be able to find more information in the "codebook" that SNOMED has published on these mappings. That codebook is named "doc_Icd10cmMapReleaseNotes_Current-en-US_US1000124_20210901.pdf" which I have zipped up for you (w/ some other related files) here: 
[SNOMED ICD10CM map technical guide etc.zip](https://github.com/mapping-commons/sssom-py/files/9122043/SNOMED.ICD10CM.map.technical.guide.etc.zip)
* test: Sure thing, I'll add one.

joeflack4 replied:
@joeflack4 TODO: Note to self: Add a test that does `validate_file` on the output.

joeflack4 commented:
> I cannot review this with these spurious diffs.. Can we somehow avoid these?

@matentzn Just to clarify, by spurious, do you mean the review comments that are linked to code snippets that are marked with the `Outdated` label?

I don't think that's really something that can easily be changed. The way to change that would be for me to write code once and for it to be perfect and then thus not have the need to change it. But I don't really think that's realistic. I think it's probably better to hack away at things fast, get something pushed, and then make future edits as needed; agile style.

joeflack4 commented:
@hrshdhgd Thanks so much for addressing the style / tox related changes. @matentzn @hrshdhgd honestly I totally forgot to run tox and make sure my code was compliant before re-pushing. Will eventually remember to do that every time, just need to make a habit.

joeflack4 commented:
[@matentzn wrote](https://github.com/mapping-commons/sssom-py/pull/207#pullrequestreview-899968455):
> I am super excited about this, and I am very happy about how much care you are taking sorting out the mapping here. Most of the things in here we need to discuss in person, too much to unpack to type up now, but two things would be probably good:
> 
> 1. Start a new "parser" documentation as described in comments above
> 2. Rebrand the method to "snomed_map" rather than "snomed_icd10_map". In fact, there are two very different ones: complex and simple map. So lets figure out which of the two you are dealing with at the moment, see also: https://github.com/mapping-commons/sssom/issues/42 and rebrand your methods accordingly. So `read_snomed_simple_map()` or `read_snomed_complex_map()` depending on what we are dealing with here.


1. Yep, will do.
2.a. Simple vs Complex
You must be looking at some old code. I remember we discussed this back in March I think, and I investigated and determined that I am using the complex map. I updated my code a few months ago; all of the methods and CLI related to this feature are named "complex". I have not utilized the simple map yet (not sure where to download it either), and that is not on my radar at the moment. I think we should probably merge this feature with only the complex map implementation, and we can always add simple later.
2.b. SNOMED map vs SNOMED::ICD map
Ok, I can rename this. But just to clarify: Are you sure that SNOMED publishes other mappings that are in the same format as the SNOMED::ICD mappings? If not, then we should continue calling this "snomed_icd_map". But if they are in the same format, then yeah I agree renaming to just "snomed_map". I can look into this for you, but I don't want it to necessarily slow down this PR. The longer we wait to merge, the more time I'll have to spend rebasing my code.

cthoyt commented:
Is there interest to finish this?

joeflack4 commented:
@cthoyt I hate to see unfinished work. I think I had basically finished "complex map" or "simple map". Other priorities took hold and this was forgotten. I leave it to @matentzn to decide, but perhaps the best thing to do is (i) close this PR, (ii) open an issue about finishing this, linking to this PR.

matentzn commented:
@joeflack4 ok! Feel free to do
As you see fit :)

joeflack4 commented:
@Nico OK then. The issue for this is still open: 
- #208 

I just updated it a little bit.

@cthoyt I could see this PR staying open for years / never getting done. So if you'd like to close it, feel free, and we can keep the above issue open.

matentzn commented:
Let leave in draft stage, I can see this becoming important in mid 2024
